dcicutils/project_utils.py

```diff
@@ -0,0 +1,453 @@
+                             f" because {cls.__name__}.declare_project_registry_class(...) has not been done.")
+        return registry_class
+
+    @classmethod
+    def declare_project_registry_class(cls, registry_class: Type):
+        """
+        This function is needed to complete a circularity where project classes need to know the appropriate
+        registry class and registry classes need to know the appropriate project class, so by convention when
+        the registry class (which is defined after the project class) is fully defined, this method on the project
+        class should be called with the registry class.
+        """
+        if not isinstance(registry_class, type) or not issubclass(registry_class, ProjectRegistry):
+            raise ValueError(f"The registry_class, {registry_class!r}, is not a subclass of ProjectRegistry.")
+        registry_class: Type[ProjectRegistry]
+        cls._PROJECT_REGISTRY_CLASS = registry_class
+
+    def __init__(self):
+        self._identity: ProjectIdentity = self.IDENTITY_CLASS(**self.IDENTITY)
+
+    def __str__(self):
+        return f"<{self.__class__.__name__} {id(self):x}>"
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}()"
+
+    @property
+    def identity(self) -> ProjectIdentity:
+        if self._identity is None:
+            raise ValueError(f"<{self.__class__.__name__}> failed to initialize correctly.")
+        return self._identity
+
+    @property
+    def PYPROJECT_NAME(self) -> str:
+        return self.identity.PYPROJECT_NAME
+
+    @property
+    def NAME(self):
+        return self.identity.NAME
+
+    @property
+    def REPO_NAME(self) -> str:
+        return self.identity.REPO_NAME
+
+    @property
+    def PYPI_NAME(self):
+        return self.identity.PYPI_NAME
+
+    @property
+    def PACKAGE_NAME(self) -> str:
+        return self.identity.PACKAGE_NAME
+
+    @property
+    def PRETTY_NAME(self):
+        return self.identity.PRETTY_NAME
+
+    @property
+    def APP_NAME(self):
+        return self.identity.APP_NAME
+
+    @property
+    def APP_PRETTY_NAME(self):
+        return self.identity.APP_PRETTY_NAME
+
+    @classproperty
+    def app_project(cls):  # noQA - PyCharm wants the variable name to be self
+        """
+        Project.app_project returns the actual instantiated project for app-specific behavior,
+        which might be of this class or one of its subclasses.
+
+        This access will fail if the project has not been initialized.
+        """
+        # This is a shared cell, so it doesn't matter whether we read this from ProjectRegistry
+        # or one of its subclasses. -kmp 19-May-2023
+        return ProjectRegistry.app_project
+
+    @classmethod
+    def initialize_app_project(cls, initialize_env_utils=True):
+        if initialize_env_utils:
+            EnvUtils.init()
+        project: Project = cls.PROJECT_REGISTRY_CLASS.initialize()
+        return project
```

Comment:
Will the `initialize` call here throw exception if env utils is not initialized?

Reply:
No. I don't think anything in here really ended up caring if EnvUtils got initialized or not. I can experiment with dropping the EnvUtils relation. There probably does want to be a place where this stuff all happens at once, but they are really separate facilities that should perhaps be managed separately.

Reply:
The EnvUtils stuff got removed by [layered PR #261](https://github.com/4dn-dcic/utils/pull/261), which is now merged in here. It's now totally unrelated, neither affecting nor affected by this work.